File: src/baekjoon/DP/9095(1,2,3 더하기)/solution.cpp
#include <string.h>
#include <map>
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>

#define MAXIMUM 1000000
using namespace std;

int funct(int number);
int dptable[12];

int main() {
	vector<int> vec;

	int count;
	int number;
	cin >> count;
	
	for (int i = 0; i < count; i++) {
		cin >> number;
		vec.push_back(number);
	}

	for (int i = 0; i < count; i++) {
		cout << funct(vec[i]) << endl;
		memset(dptable, 0, 12);
	}

	return 0;
}


int funct(int number) {
	
	if (number < 0) return 0;
	if (number == 0) return 1;
	if (dptable[number] > 0) return dptable[number];

	dptable[number] = funct(number - 1) + funct(number - 2) + funct(number - 3);

	return dptable[number];

}

File: src/baekjoon/DP/9095(1,2,3 더하기)/solution.py
def funct(n):
  if n < 0:
    return 0
  elif n == 0:
    return 1
  elif dptable[n] > 0:
    return dptable[n]
  
  dptable[n] = funct(n-1) + funct(n-2) + funct(n-3)

  return dptable[n]


t = int(input())

for _ in range(t):
  n = int(input())
  dptable = [0]*(n+1)
  print(funct(n))
